lib/plugins/logviewer/admin.php

```diff
@@ -153,38 +126,57 @@ protected function getFacilities()
         return $facilities;
     }
 
-    private function readEndOfLogFile($logfile, $numberOfLinesLimit)
+    /**
+     * Read the lines of the logfile and return them as an array
+     *
+     * @param string $logfilePath
+     * @return array
+     */
+    protected function getLogLines($logfilePath)
     {
         $lines = [];
-        $fp = fopen($logfile, 'r');
+        $size = filesize($logfilePath);
         
-        fseek($fp, -1, SEEK_END);
-        if(fgetc($fp) == PHP_EOL) { // if the last line is EOL ignore it
-            $pos = -2;
-        } else { // otherwise start from the last character
-            $pos = -1;
-        }
-
-        $currentLine = '';
-
-        while (count($lines) < $numberOfLinesLimit && (-1 !== fseek($fp, $pos, SEEK_END))) {
-            $char = fgetc($fp);
-            if (PHP_EOL == $char) {
-                $lines[] = $currentLine;
-                $currentLine = '';
-            } else {
-                $currentLine = $char . $currentLine;
+        if ($size <= self::MAX_READ_SIZE) {
+            $lines = file($logfilePath);
+        } else {
+            $fp = fopen($logfilePath, 'r');
+            if ($fp !== null) {
+                if (fseek($fp, -self::MAX_READ_SIZE, SEEK_END) == 0) {
+                    if (false !== ($d = fread($fp, self::MAX_READ_SIZE))) {
+                        $lines = explode("\n", $d);  // Use the same line delimiter as in
+                                                     // inc/Logger.php
+                        $d = null;  // Free up space.
+
+                        // Discard the first line as it is probably truncated:
+                        array_shift($lines);
+
+                        // Discard indented lines at the start:
+                        while (!empty($lines) && (substr($lines[0], 0, 2) === '  '))
+                            array_shift($lines);
+
+                        // Add a note (as a fake log entry) to indicate that the first
+                        // log lines in the file are not shown.
+                        // The following should probably use a localizable text?
+                        $linesSkippedMsg = 'Log file too large. Previous lines skipped!';
+                        array_unshift($lines, "******\t"."\t".'['.$linesSkippedMsg.']');
+                    }
+                }
+                fclose($fp);
             }
-            $pos--;
         }
 
-        fclose($fp);
-        return array_reverse($lines);
+        return $lines;
     }
 
-    private function printLogLines($lines)
+    /**
+     * Get an array of log lines and print them using appropriate styles 
+     * The default value is 1500 lines
+     * @return array
+     */
+    protected function printLogLines($lines, $numberOfLines = 1500)
     {
-        for ($i = 0; $i < count($lines); $i++) {
+        for ($i = 0; $i < $numberOfLines; $i++) {
```

Comment:
So what will happen when `$lines` has less than `$numberOfLines` elements? You will use an out of bounds index!

You need to take the minimum of `count($lines)` and `$numberOfLines` and loop over that number.

That said, why are you limiting the number of lines at all? There are two separate issues here:

A) The real limit is already imposed by `MAX_READ_SIZE`. I think the shortest possible log line would contain ≈22 bytes (timestamp, 2 HTABs, LF) so the absolute worst case would be ≈47700 lines. A lot sure, but not really a problem. Most lines will be significantly longer than 22 bytes so this shouldn't be an issue. Also while you have the `$numberOfLines` parameter, it will always have the default value of 1500. There is no mechanism for setting a different value.

B) The way you limit the number of lines you are getting the first 1500 lines, not the last. So given e.g. 5000 log lines, you would be showing lines 0-1499, but not showing the all important last lines 3500-4999 of the log. If the whole log file was bigger than `MAX_READ_SIZE` then you are displaying some arbitrary lines from the middle of the log — without telling the reader about it to boot.

And once you get rid of the line number limit, there is no need to refactor the line display code into its own method anymore. Just keep the original code as it was.

If you do want to refactor the code, arguably the `<dl> … </dl>` parts should be part of the `printLogLines()` method, not outside of it. A different implementation might want to use a `<table>` or just a bunch of `<div>` elements as containers for a line.

Reply:
I made a simple change, It will read the last 1mb, and `count` how many lines it is, and then show it (no matter how many is that 1mb). No problem anymore.